sound/soc/sof/ipc4-topology.c

```diff
@@ -1399,6 +1479,86 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 	return 0;
 }
 
+static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
+					   struct snd_pcm_hw_params *fe_params,
+					   struct snd_sof_platform_stream_params *platform_params,
+					   struct snd_pcm_hw_params *pipeline_params, int dir)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_soc_dapm_widget *widget = swidget->widget;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct sof_ipc4_process *process = swidget->private;
+	struct snd_sof_control *scontrol = NULL;
+	const struct snd_kcontrol_new *kc;
+	struct soc_bytes_ext *sbe;
+	struct sof_ipc4_control_data *control_data;
+	struct sof_ipc4_available_audio_format *available_fmt = &process->available_fmt;
+	void *cfg = process->ipc_config_data;
+	void *extra_flag;
+	int ret, i;
+
+	available_fmt->ref_audio_fmt = &available_fmt->base_config->audio_fmt;
+
+	/*
+	 * Output format is optional for process modules.
+	 * Process modules setup the output format based on audio format tokens in topology.
+	 */
+	if (swidget->output_fmt_cnt)
+		ret = sof_ipc4_init_audio_fmt(sdev, swidget, &process->base_config,
+					      &process->output_format, pipeline_params,
+					      available_fmt, sizeof(process->base_config));
+	else
+		ret = sof_ipc4_init_audio_fmt(sdev, swidget, &process->base_config,
+					      NULL, pipeline_params, available_fmt,
+					      sizeof(process->base_config));
+	if (ret < 0)
+		return ret;
+
+	/* update pipeline memory usage */
+	sof_ipc4_update_pipeline_mem_usage(sdev, swidget, &process->base_config);
+
+	/*
+	 * ipc_config_data is composed of the base_config, optional output formats followed
+	 * by the data required for module init in that order.
+	 */
+	memcpy(cfg, &process->base_config, sizeof(process->base_config));
+	cfg += sizeof(process->base_config);
+
+	/* copy output_format to configure data payload */
+	for (i = 0; i < swidget->output_fmt_cnt; i++) {
+		/* TODO: to handle different output formats */
+		memcpy(cfg, &process->output_format, sizeof(struct sof_ipc4_audio_format));
+		cfg += sizeof(struct sof_ipc4_audio_format);
+	}
+
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+		if (!kc)
+			return -ENODEV;
+
+		/* payload uses byte kcontrol */
+		if (widget->dobj.widget.kcontrol_type[i] != SND_SOC_TPLG_TYPE_BYTES)
+			continue;
+		sbe = (struct soc_bytes_ext *)kc->private_value;
+		scontrol = sbe->dobj.private;
+		if (!scontrol)
+			continue;
+		control_data = scontrol->ipc_control_data;
+		if (control_data->data->ipc_blob_type != SOF_IPC4_MOD_INIT_INSTANCE)
+			continue;
+
+		/*
+		 * TODO: add flag in topology to indicate whether the module has
+		 * SOF_IPC4_MOD_INIT_INSTANCE type kcontrol and do sanity check here
```

Comment:
what sanity test we want to do here?

Reply:
@aiChaoSONG Pierre and Ranjani suggested to add a flag to indicate whether the module has init_instance and/or (large_)config or not.

Reply:
@ranj063 I don't get the idea why we should have tokens like SOF_IPC4_MOD_INIT_INSTANCE to tell us if there is kcontrol for module init, which we can do by iterating the kcontrol list, and check the ipc_blob_type in the header.

Reply:
Yes.

Reply:
Is this comment still valid and if it is why it is after you have checked the blob type against SOF_IPC4_MOD_INIT_INSTANCE?

Reply:
Yes, this comment is still valid. It's a feature. As @ranj063 suggested I will implement for saint check. I will set a flag in tplg to imply whether the module instance has blob for init instance and/or config/large_config. And the driver will did the check.